File: JbmrksIOs/JbmrksIOs/ViewModels/TaskFormViewModel.swift
//
//  TaskFormViewModel.swift
//  JbmrksIOs
//
//  ViewModel for Task Form screen
//

import Foundation
import Combine

@MainActor
final class TaskFormViewModel: ObservableObject {
    @Published var title: String = ""
    @Published var description: String = ""
    @Published var deadline: String? = nil
    @Published var priority: TaskPriority = .normal
    @Published var isLoading = false
    @Published var isSaving = false
    @Published var errorMessage: String?
    
    private var tasksRepository: TasksRepository?
    private let taskId: String?
    
    var isCreating: Bool {
        taskId == nil || taskId == "new"
    }
    
    init(taskId: String? = nil) {
        self.taskId = taskId
    }
    
    func loadTask() async {
        guard let taskId = taskId, !isCreating else {
            // Creating new task - no need to load
            isLoading = false
            return
        }
        
        isLoading = true
        errorMessage = nil
        tasksRepository = RepositoryFactory.shared.tasksRepository()
        
        guard let repo = tasksRepository else {
            errorMessage = "Not authenticated"
            isLoading = false
            return
        }
        
        do {
            let task = try await repo.getTask(id: taskId)
            title = task.title
            description = task.description
            deadline = task.deadline
            priority = task.priority
        } catch {
            errorMessage = "Failed to load task: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    func saveTask() async throws {
        guard let repo = tasksRepository ?? RepositoryFactory.shared.tasksRepository() else {
            throw NSError(domain: "TaskFormViewModel", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not authenticated"])
        }
        
        tasksRepository = repo
        
        guard !title.isEmpty else {
            errorMessage = "Title is required"
            throw NSError(domain: "TaskFormViewModel", code: -1, userInfo: [NSLocalizedDescriptionKey: "Title is required"])
        }
        
        isSaving = true
        errorMessage = nil
        
        do {
            if isCreating {
                _ = try await repo.createTask(
                    title: title,
                    description: description.isEmpty ? nil : description,
                    deadline: deadline,
                    priority: priority
                )
            } else {
                guard let taskId = taskId else {
                    throw NSError(domain: "TaskFormViewModel", code: -1, userInfo: [NSLocalizedDescriptionKey: "Task ID required for update"])
                }
                _ = try await repo.updateTask(
                    id: taskId,
                    title: title,
                    description: description.isEmpty ? nil : description,
                    deadline: deadline,
                    priority: priority
                )
            }
        } catch {
            errorMessage = isCreating ? "Failed to create task: \(error.localizedDescription)" : "Failed to update task: \(error.localizedDescription)"
            isSaving = false
            throw error
        }
        
        isSaving = false
    }
    
    // Legacy method for backward compatibility
    func updateTask() async throws {
        try await saveTask()
    }
}

File: app/src/main/java/com/example/jbmarks/tasks/ui/TaskFormViewModel.kt
package com.example.jbmarks.tasks.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.jbmarks.tasks.data.TasksRepository
import com.example.jbmarks.tasks.domain.TaskPriority
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

sealed interface TaskFormUiState {
    object Loading : TaskFormUiState
    data class Editing(
        val title: String = "",
        val description: String = "",
        val deadline: String? = null,
        val priority: TaskPriority = TaskPriority.NORMAL,
        val isSaving: Boolean = false,
        val error: String? = null
    ) : TaskFormUiState
    object Saved : TaskFormUiState
}

class TaskFormViewModel(
    private val taskId: String?,
    private val repository: TasksRepository = TasksRepository()
) : ViewModel() {

    private val _uiState = MutableStateFlow<TaskFormUiState>(
        if (taskId == null) {
            TaskFormUiState.Editing()
        } else {
            TaskFormUiState.Loading
        }
    )
    val uiState: StateFlow<TaskFormUiState> = _uiState

    fun loadTask(taskId: String) {
        viewModelScope.launch {
            _uiState.value = TaskFormUiState.Loading
            repository.getTask(taskId)
                .onSuccess { task ->
                    _uiState.value = TaskFormUiState.Editing(
                        title = task.title,
                        description = task.description,
                        deadline = task.deadline,
                        priority = task.priority
                    )
                }
                .onFailure { throwable ->
                    _uiState.value = TaskFormUiState.Editing(
                        error = throwable.message ?: "Failed to load task"
                    )
                }
        }
    }

    fun updateTitle(title: String) {
        val currentState = _uiState.value
        if (currentState is TaskFormUiState.Editing) {
            _uiState.value = currentState.copy(title = title, error = null)
        }
    }

    fun updateDescription(description: String) {
        val currentState = _uiState.value
        if (currentState is TaskFormUiState.Editing) {
            _uiState.value = currentState.copy(description = description, error = null)
        }
    }

    fun updateDeadline(deadline: String?) {
        val currentState = _uiState.value
        if (currentState is TaskFormUiState.Editing) {
            _uiState.value = currentState.copy(deadline = deadline, error = null)
        }
    }

    fun updatePriority(priority: TaskPriority) {
        val currentState = _uiState.value
        if (currentState is TaskFormUiState.Editing) {
            _uiState.value = currentState.copy(priority = priority, error = null)
        }
    }

    fun createTask(onSuccess: () -> Unit) {
        val currentState = _uiState.value
        if (currentState !is TaskFormUiState.Editing) return

        if (currentState.title.isBlank()) {
            _uiState.value = currentState.copy(error = "Title is required")
            return
        }

        viewModelScope.launch {
            _uiState.value = currentState.copy(isSaving = true, error = null)
            
            repository.createTask(
                title = currentState.title,
                description = currentState.description.takeIf { it.isNotBlank() },
                deadline = currentState.deadline,
                priority = currentState.priority
            )
                .onSuccess {
                    android.util.Log.d("TaskFormViewModel", "Task created successfully, navigating back")
                    _uiState.value = TaskFormUiState.Saved
                    onSuccess()
                }
                .onFailure { throwable ->
                    _uiState.value = currentState.copy(
                        isSaving = false,
                        error = throwable.message ?: "Failed to create task"
                    )
                }
        }
    }

    fun updateTask(taskId: String, onSuccess: () -> Unit) {
        val currentState = _uiState.value
        if (currentState !is TaskFormUiState.Editing) return

        if (currentState.title.isBlank()) {
            _uiState.value = currentState.copy(error = "Title is required")
            return
        }

        viewModelScope.launch {
            _uiState.value = currentState.copy(isSaving = true, error = null)
            
            repository.updateTask(
                taskId = taskId,
                title = currentState.title,
                description = currentState.description.takeIf { it.isNotBlank() },
                deadline = currentState.deadline,
                priority = currentState.priority
            )
                .onSuccess {
                    android.util.Log.d("TaskFormViewModel", "Task updated successfully, navigating back")
                    _uiState.value = TaskFormUiState.Saved
                    onSuccess()
                }
                .onFailure { throwable ->
                    _uiState.value = currentState.copy(
                        isSaving = false,
                        error = throwable.message ?: "Failed to update task"
                    )
                }
        }
    }
}

class TaskFormViewModelFactory(
    private val taskId: String?
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(TaskFormViewModel::class.java)) {
            return TaskFormViewModel(taskId) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
